plugins/modem-manager/fu-mm-device.c

```diff
@@ -15,8 +15,12 @@
 #include "fu-mm-utils.h"
 #include "fu-qmi-pdc-updater.h"
 
-/* amount of time for the modem to be re-probed and exposed in MM after being uninhibited */
-#define FU_MM_DEVICE_REMOVE_DELAY_REPROBE	45000	/* ms */
+/* Amount of time for the modem to be re-probed and exposed in MM after being
+ * uninhibited. The timeout is long enough to cover the worst case, where the
+ * modem boots without SIM card inserted (and therefore the initialization
+ * may be very slow) and also where carrier config switching is explicitly
+ * required (e.g. if switching from the default (DF) to generic (GC).*/
+#define FU_MM_DEVICE_REMOVE_DELAY_REPROBE	120000	/* ms */
```

Comment:
That's insanely long.  Have you checked to make sure that the frontend dbus connection doesn't time out during this?

Reply:
I know it's long yes... but I've seen modems needing up to 45s in a "normal boot" until they're functional, and then the actual ModemManager port probing may take a long time under several cases (e.g. if there is no SIM detected, we retry several times because the SIM initialization also takes its own time). Plus, the first probing after a firmware upgrade may end up triggering a carrier config switching in ModemManager, which will reboot the device again... I'm just looking at the worst case scenario here, the timeout is just long enough to cover that. Usually, the actual timeout may be a few seconds.

There's no timeout in DBus because this is not a method for which we require response; during that wait time we're just waiting for the module to appear exposed in ModemManager.

Reply:
OK thanks.

Reply:
For some context regarding timing, the full DW5821e firmware upgrade process, including the fastboot and QMI PDC uploads, plus the carrier config selection, is right now taking 4 minutes 15s (with and without SIM).

The real time needed for the reattach after the upgrade in that scenario is around 35s (not 120s):

```
13:19:32:0848 FuPlugin             emit removed from modem_manager: 57739e832c92f1f3822041764876111358d2eb83
13:19:32:0848 FuDeviceList         waiting 120000ms for DW5821e Snapdragon X20 LTE device removal
13:19:32:0868 FuEngine             auto-removing GUsbDevice
13:19:32:0869 FuDeviceList         waiting 120000ms for DW5821e Snapdragon X20 LTE device removal
13:19:33:0955 FuEngine             no plugin specified for USB device 413c:81d7
13:19:34:0117 FuEngine             auto-removing GUsbDevice
13:19:34:0117 FuDeviceList         waiting 120000ms for DW5821e Snapdragon X20 LTE device removal
13:19:36:0112 FuPluginMm           request to activate configuration timed out: assuming device reset is ongoing
13:19:36:0112 Qmi                  [/dev/cdc-wdm0] Releasing 'pdc' client with flags 'release-cid'...
13:19:36:0112 Qmi                  [/dev/cdc-wdm0] Unregistered 'pdc' client with ID '1'
13:19:36:0112 FuPluginMm           qmi-pdc attach operation successful
13:19:36:0113 FuPluginMm           uninhibit modemmanager device with uid /sys/devices/pci0000:00/0000:00:1d.0/usb2/2-1/2-1.4/2-1.4.1
13:19:42:0717 FuEngine             no plugin specified for USB device 413c:81d7
13:20:07:0982 FuDevice             using 57739e832c92f1f3822041764876111358d2eb83 for /sys/devices/pci0000:00/0000:00:1d.0/usb2/2-1/2-1.4/2-1.4.1
13:20:07:0982 FuPlugin             emit added from modem_manager: 57739e832c92f1f3822041764876111358d2eb83
```

Reply:
> 4 minutes 15s

Can we provide any information back to the user in this case? e.g. a percentage bar? After a couple minutes of inactivity, even I'm going to be rebooting as the upgrade "failed".

Reply:
hum... we already have the progress bar reported by fwupdmgr, don't we?

Reply:
Is it just going backwards and forwards in that case for 5 minutes?

Reply:
The only "a bit confusing" output from fwupdmgr is when the second pass to run QMI PDC is done after having done the fastboot update; but the progress never goes "backwards". Improving the progress reported was already in my pending task list, truth be told.

Note anyway, that this MR won't really make the process slower, it will actually make it a bit faster if anything, because the carrier selection is much quicker when done by fwupd than when done by MM (fwupd knows what the "last" used carrier was, while MM would need to query the SIM card).

The max 120s new threshold is just a "safe enough" value, and it was only being >45s for a very specific case when MM was the one trying to do the carrier config selection. With the additional patch to do the carrier config selection by fwupd itself, the real time needed to re-detect the modem by MM after the full process has finished is around 35s as per the logs in a previous comment in this thread. So in reality we would even be fine for the current usecase even with the 45->120 upgrade, but I thought that leaving it wouldn't be a big problem.

As you noted, the key point is to have the progress reporting right I think. Plus, I believe I could make the qmi-pdc loading a bit faster, but that's unrelated.